src/network/networkd-radv.c

```diff
@@ -21,9 +21,138 @@
 #include <arpa/inet.h>
 
 #include "networkd-address.h"
+#include "networkd-manager.h"
 #include "networkd-radv.h"
 #include "sd-radv.h"
 
+static int radv_get_ip6dns(Network *network, struct in6_addr **dns,
+                           size_t *n_dns) {
+        _cleanup_free_ struct in6_addr *addresses = NULL;
+        size_t i, n_addresses = 0, n_allocated = 0;
+
+        assert(network);
+        assert(dns);
+        assert(n_dns);
+
+        for (i = 0; i < network->n_dns; i++) {
+                union in_addr_union addr;
+
+                if (network->dns[i].family != AF_INET6)
+                        continue;
+
+                addr.in6 = network->dns[i].address.in6;
+
+                if (in_addr_is_null(AF_INET6, &addr) ||
+                    in_addr_is_link_local(AF_INET6, &addr) ||
+                    in_addr_is_localhost(AF_INET6, &addr))
+                        continue;
+
+                if (!GREEDY_REALLOC(addresses, n_allocated, n_addresses + 1))
+                        return log_oom();
+
+                addresses[n_addresses++] = addr.in6;
+        }
+
+        if (addresses) {
+                free(*dns);
+                *dns = addresses;
```

Comment:
hmm, normally we prefer that functions have parameters that are either input or output, but not both, except if they refer to proper "objects" with a naturally longer lifetime. Exceptions are OK to have, if there's a good reason to, but they should probably have a comment, clarifying that the parameters are both input and output, like in this case

Reply:
Fixed not free anything anymore, just sets the pointer. Freeing was anyway an unnecessary operation even from the point of the calling code.